lighthouse-viewer/app/src/main.js

```diff
@@ -0,0 +1,518 @@
+          date: new Date(jsonFile.generatedTime)
+        });
+        const body = {
+          description: 'Lighthouse json report',
+          public: false,
+          files: {
+            [`${filename}${GithubAPI.LH_JSON_EXT}`]: {
+              content: JSON.stringify(jsonFile)
+            }
+          }
+        };
+
+        const request = new Request('https://api.github.com/gists', {
+          method: 'POST',
+          headers: new Headers({Authorization: `token ${accessToken}`}),
+          // Stringify twice so quotes are escaped for POST request to succeed.
+          body: JSON.stringify(body)
+        });
+        return fetch(request);
+      })
+      .then(resp => resp.json())
+      .then(json => {
+        logger.hide();
+        return json.id;
+      });
+  }
+
+  /**
+   * Fetches the body content of a gist.
+   * @param {!string} id The id of a gist.
+   * @return {!Promise<{etag: string, content: !Object}>} json content of the gist.
+   */
+  getGistContent(id) {
+    return this.auth.ready.then(user => {
+      const headers = new Headers();
+
+      // If there's an authenticated user, include an Authorization header to
+      // have higher rate limits with the Github API. Otherwise, rely on ETags.
+      if (user) {
+        headers.set('Authorization', `token ${this.auth.accessToken}`);
+      }
+
+      return idb.get(id).then(cachedGist => {
+        if (cachedGist && cachedGist.etag) {
+          headers.set('If-None-Match', cachedGist.etag);
+        }
+
+        // Always make the request to see if there's newer content.
+        return fetch(`https://api.github.com/gists/${id}`, {headers}).then(resp => {
+          const remaining = resp.headers.get('X-RateLimit-Remaining');
+          const limit = resp.headers.get('X-RateLimit-Limit');
+          if (Number(remaining) < 10) {
+            logger.warn('Warning: approaching Github\'s rate limit. ' +
+                        `${limit - remaining}/${limit} requests used. Consider signing ` +
+                        'in to increase this limit.');
+          }
+
+          const etag = resp.headers.get('ETag');
+
+          if (!resp.ok) {
+            if (resp.status === 304) {
+              return cachedGist;
+            } else if (resp.status === 404) {
+              // Delete the entry from IDB if it no longer exists on the server.
+              idb.delete(id); // Note: async.
+            }
+            const msg = `${resp.status} fetching gist`;
+            logger.log(msg);
+            throw new Error(msg);
+          }
+
+          return resp.json().then(json => {
+            const fileName = Object.keys(json.files)[0]; // Attempt to use first file in gist.
+            const f = json.files[fileName];
+            if (f.truncated) {
+              return fetch(f.raw_url).then(resp => resp.json())
+                  .then(json => ({etag, content: json}));
+            }
+            return {etag, content: JSON.parse(f.content)};
```

Comment:
parsing here keeps the promise's resolve value consistent with the truncated case (where we fetch the full file and `resp.json()`).